### PR TITLE
add resource based scheduling prioritizer

### PR DESCRIPTION
**Pull request by haoqing0110**
Ref: https://github.com/open-cluster-management/backlog/issues/15599

Signed-off-by: haoqing0110 <qhao@redhat.com>

### CODE REVIEW

File: pkg/controllers/scheduling/schedule.go [hunk 1]
```diff
@@ -113,13 +114,22 @@ type pluginScheduler struct {
 }
 
 func NewPluginScheduler(handle plugins.Handle) *pluginScheduler {
+	resourceAllocatableCPU := &resource.ResourceAllocatableCPU{}
```
Comment (qiujian16) [inline pkg/controllers/scheduling/schedule.go:117]: I think we can make the code more general here. For example we could have a ResourcePrioritizerFactory. So you will do like

```
resourcePrioritizerFactory := resource.NewPrioritizerFactory(handle)

resourcePrioritizerFactory.Resource("cpu").Mode("Allocatable").Build()
resourcePrioritizerFactory.Resource("memory").Mode("Allocatable").Build()
```
And you would not need to create 4 plugins, but a more general one. WDYT?

Comment (qiujian16) [reply pkg/controllers/scheduling/schedule.go:117]: Or we might want a dynamic generated prioritizer lists.

```go
prioritizerFactory := prioritizer.NewFactory(handle)
...

// In scheduler we build prioritizers for each placement
prioritizers := priotizerFactory.WithPlacement(placement).Build()
```
We can make it as a followup.

File: pkg/plugins/resource/resource.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	"fmt"
+	"sort"
+
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:14]: nit: you can use clusterv1alaph1.PlacementLabel now.

File: pkg/plugins/resource/resource.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	The clusters that has the most allocatable to capacity ratio or allocatable are given the highest score, 
+	while the least is given the lowest score.
+	`
+)
+
+var _ plugins.Prioritizer = &Resource{}
+
+type Resource struct {
+	handle   plugins.Handle
+	resource string
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:28]: clusterapiv1.ResourceName?

File: pkg/plugins/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	mode     string
+}
+
+type ResourceBuilder struct {
+	resource *Resource
+}
+
+func NewResourceBuilder(handle plugins.Handle) *ResourceBuilder {
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:36]: how about NewResourcePrioritizerBuilder(handle).WithPrioritizerName("resourceCPUAllocatable").Build()

and you can maintain a map here
```go
resourceMap := map[string]clusterapiv1.ResourceName {
    "CPU":  clusterapiv1.ResourceCPU
    ...
}

func (r *ResourcePrioritizerBuilder) Build() {
    resource, mode := parsePrioritizerName(r.prioritizerName)
    return &resourcePrioritizer{
        resource: resourceMap[resource], // or this map find is put in parsePrioritizerName
        mode: mode,
        name: r.prioritizerName,
    }
}
```

File: pkg/plugins/resource/resource.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+func (r *Resource) Description() string {
+	return description
+}
+
+func (r *Resource) Score(ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error) {
+	f := r.mode + r.resource
+	switch {
+	case f == "RatioMemory":
```
Comment (qiujian16) [inline pkg/plugins/resource/resource.go:69]: with above thing, what you need to do here is

```
switch mode {
case "Ratio":
    return mostResourceRatioScores(c.resource, clusters)
case "Allocatable":
    return mostResourceAllocatableScores(c.resource, clusters)
}
```

Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: Updated.

File: pkg/plugins/resource/resource.go (hunk 4 above)
Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: Updated.

File: pkg/plugins/resource/resource.go (hunk 3 above)
Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: Updated.

File: pkg/plugins/resource/resource.go (hunk 2 above)
Comment (haoqing0110) [reply pkg/plugins/resource/resource.go]: Updated.

**Review by qiujian16 — COMMENTED** [review]
/approve

**Pull request merged**